### PR TITLE
GH-41813: [C++] Fix avx2 gather offset larger than 2GB in `CompareColumnsToRows`

### PR DESCRIPTION
**Pull request by zanmato1984**
### Rationale for this change

AVX2 intrinsics `_mm256_i32gather_epi32`/`_mm256_i32gather_epi64` are used in `CompareColumnsToRows` API, and treat the `vindex` as signed integer. In our row table implementation, we use `uint32_t` to represent the offset within the row table. When a offset is larger than (`0x80000000`, or `2GB`), the aforementioned intrinsics will treat it as negative offset and gather the data from undesired address. More details please see https://github.com/apache/arrow/issues/41813#issuecomment-2168286330.

Considering there is no unsigned-32bit-offset or 64bit-offset counterparts of those intrinsics in AVX2, this issue can be simply mitigated by translating the base address and the offset:
```
new_base = base + 0x80000000;
new_offset = offset - 0x80000000;
```

### What changes are included in this PR?

Fix and UT that reproduces the issue.

### Are these changes tested?

UT included.

### Are there any user-facing changes?

None.

* GitHub Issue: #41813

### CODE REVIEW

**Comment by zanmato1984** [comment]
cc @pitrou @amoeba @mrd0ll4r

File: cpp/src/arrow/compute/row/compare_test.cc [hunk 1]
```diff
@@ -164,5 +166,126 @@ TEST(KeyCompare, CompareColumnsToRowsTempStackUsage) {
...
+  }
+
+  // The idea of this case is to create a row table using several fixed length columns and
+  // one var length column (so the row is hence var length and has offset buffer), with
+  // the overall data size exceeding 2GB. Then compare each row with itself.
+  constexpr int64_t two_gb = 2ll * 1024ll * 1024ll * 1024ll;
+  // The compare function requires the row id of the left column to be uint16_t, hence the
+  // number of rows.
+  constexpr int64_t num_rows = std::numeric_limits<uint16_t>::max() + 1;
+  // TODO: This test will fail if we switch the order between uint64 and uint32.
```
Comment (zanmato1984) [inline cpp/src/arrow/compute/row/compare_test.cc:183]: This is discovered by accident, and I'm positive that this is non-related to this fix. So I intended to do this fix first and revisit this TODO in the near future.

**Comment by amoeba** [comment]
Hi @zanmato1984, thanks for your work on this. I'm hoping others can review the implementation but I did just check that the new test passes (it does) and also fixes the original issue (it does). :+1:

**Comment by zanmato1984** [comment]
> Hi @zanmato1984, thanks for your work on this. I'm hoping others can review the implementation but I did just check that the new test passes (it does) and also fixes the original issue (it does). 👍

Thank you @amoeba for verifying, and the help on reproducing the issue!

**Comment by mrd0ll4r** [comment]
I can't give much feedback or test this out, unfortunately. But I'm very thankful for you all looking into this!

**Comment by FreekPaans** [comment]
Ran into this issue when I was debugging my own issue where running a group_by/aggregate on a table with null columns was failing to group some keys, i.e. some group key value tuples were duplicated in the result. 

Why I mention it here
:
* My investigation also pointed to a problem with CompareColumnsToRows, but I don't fully understand the code.
* Disabling AVX2 (i.e. `ARROW_USER_SIMD_LEVEL=AVX) solved the issue

However, the table I use is only 3.8MB - so perhaps there is some other bug around the AVX2 related code here as well, unrelated to the size, but related to nulls.

**Comment by amoeba** [comment]
Hi @FreekPaans, can you please open a new issue for that? I think the issue will be fixed in the upcoming 17.x PyArrow release but it'd be good to make sure.

**Comment by FreekPaans** [comment]
@amoeba sure thing https://github.com/apache/arrow/issues/42231

Any issue/PR you can point me to for where it's fixed?

**Comment by zanmato1984** [comment]
@pitrou @felipecrv @ZhangHuiGui @mapleFU Would you please help to take a look? Thanks.

File: cpp/src/arrow/compute/row/compare_test.cc (hunk 1 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_test.cc]: Verified to be a mis-use. TODO removed.

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc [hunk 2]
```diff
@@ -236,6 +236,8 @@ uint32_t KeyCompare::CompareBinaryColumnToRowHelper_avx2(
         irow_right =
             _mm256_loadu_si256(reinterpret_cast<const __m256i*>(left_to_right_map) + i);
       }
+      // TODO: Need to test if this gather is OK when irow_right is larger than
```
Comment (zanmato1984) [inline cpp/src/arrow/compute/row/compare_internal_avx2.cc:239]: I'll test in the future.

Comment (pitrou) [reply cpp/src/arrow/compute/row/compare_internal_avx2.cc:239]: When you say "in the future", is it in this PR or another one?

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc [hunk 3]
```diff
@@ -251,6 +253,35 @@ uint32_t KeyCompare::CompareBinaryColumnToRowHelper_avx2(
...
+
+/// Intrinsics `_mm256_i32gather_epi32/64` treat the `vindex` as signed integer, and we
+/// are using `uint32_t` to represent the offset, in range of [0, 4G), within the row
+/// table. When the offset is larger than `0x80000000` (2GB), those intrinsics will treat
+/// it as negative offset and gather the data from undesired address. To avoid this issue,
+/// we normalize the addresses by translating `base` `0x80000000` higher, and `offset`
+/// `0x80000000` lower. This way, the offset is always in range of [-2G, 2G) and those
+/// intrinsics are safe.
+
+constexpr auto two_gb = 0x80000000ull;
```
Comment (pitrou) [inline cpp/src/arrow/compute/row/compare_internal_avx2.cc:266]: Can we make sure we use an explicit width type here? I'm not even sure what it is expected to be for correctness of the code using this constant (`uint32_t` or `uint64_t`?)

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc [hunk 4]
```diff
@@ -251,6 +253,35 @@ uint32_t KeyCompare::CompareBinaryColumnToRowHelper_avx2(
...
+
+template <int scale>
+inline __m256i UnsignedOffsetSafeGather32(int const* base, __m256i offset) {
+  auto normalized_base = base + two_gb / sizeof(int);
+  __m256i normalized_offset = _mm256_sub_epi32(offset, _mm256_set1_epi32(two_gb / scale));
+  return _mm256_i32gather_epi32(normalized_base, normalized_offset, scale);
+}
+
+template <int scale>
+inline __m256i UnsignedOffsetSafeGather64(arrow::util::int64_for_gather_t const* base,
```
Comment (pitrou) [inline cpp/src/arrow/compute/row/compare_internal_avx2.cc:276]: What is the use of `int64_for_gather_t` exactly?

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc [hunk 5]
```diff
@@ -251,6 +253,35 @@ uint32_t KeyCompare::CompareBinaryColumnToRowHelper_avx2(
...
+/// are using `uint32_t` to represent the offset, in range of [0, 4G), within the row
+/// table. When the offset is larger than `0x80000000` (2GB), those intrinsics will treat
+/// it as negative offset and gather the data from undesired address. To avoid this issue,
+/// we normalize the addresses by translating `base` `0x80000000` higher, and `offset`
+/// `0x80000000` lower. This way, the offset is always in range of [-2G, 2G) and those
+/// intrinsics are safe.
+
+constexpr auto two_gb = 0x80000000ull;
+
+template <int scale>
```
Comment (pitrou) [inline cpp/src/arrow/compute/row/compare_internal_avx2.cc:268]: Two things:
1) if we're using unsigned arithmetic below, the scale type should probably be unsigned for readability and sanity?
2) naming convention: can we make this `kScale`?

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc [hunk 6]
```diff
@@ -251,6 +253,35 @@ uint32_t KeyCompare::CompareBinaryColumnToRowHelper_avx2(
   }
 }
 
+namespace {
+
+/// Intrinsics `_mm256_i32gather_epi32/64` treat the `vindex` as signed integer, and we
```
Comment (pitrou) [inline cpp/src/arrow/compute/row/compare_internal_avx2.cc:258]: Can you use regular comments (`//`)? This isn't a docstring so shouldn't use the docstring-specific prefix (`///`)

File: cpp/src/arrow/compute/row/compare_test.cc [hunk 7]
```diff
@@ -164,5 +166,128 @@ TEST(KeyCompare, CompareColumnsToRowsTempStackUsage) {
...
+
+  // The idea of this case is to create a row table using several fixed length columns and
+  // one var length column (so the row is hence var length and has offset buffer), with
+  // the overall data size exceeding 2GB. Then compare each row with itself.
+  constexpr int64_t two_gb = 2ll * 1024ll * 1024ll * 1024ll;
+  // The compare function requires the row id of the left column to be uint16_t, hence the
+  // number of rows.
+  constexpr int64_t num_rows = std::numeric_limits<uint16_t>::max() + 1;
+  const std::vector<std::shared_ptr<DataType>> fixed_length_types{uint64(), uint32()};
+  // The var length column should be a little smaller than 2GB to WAR the capacity
```
Comment (pitrou) [inline cpp/src/arrow/compute/row/compare_test.cc:184]: "WAR"?

File: cpp/src/arrow/compute/row/compare_test.cc [hunk 8]
```diff
@@ -164,5 +166,128 @@ TEST(KeyCompare, CompareColumnsToRowsTempStackUsage) {
   }
 }
 
+// Compare columns to rows at offsets over 2GB within a row table.
+// Certain AVX2 instructions may behave unexpectedly causing troubles like GH-41813.
+TEST(KeyCompare, CompareColumnsToRowsLarge) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/row/compare_test.cc:171]: What is the runtime of this test? Perhaps we need to disable it on Valgrind builds.

File: cpp/src/arrow/compute/row/compare_test.cc [hunk 9]
```diff
@@ -164,5 +166,128 @@ TEST(KeyCompare, CompareColumnsToRowsTempStackUsage) {
...
+  ASSERT_OK(row_table.Init(pool, table_metadata_right));
+  std::vector<uint16_t> row_ids_right(num_rows);
+  std::iota(row_ids_right.begin(), row_ids_right.end(), 0);
+  RowTableEncoder row_encoder;
+  row_encoder.Init(column_metadatas, sizeof(uint64_t), sizeof(uint64_t));
+  row_encoder.PrepareEncodeSelected(0, num_rows, column_arrays);
+  ASSERT_OK(row_encoder.EncodeSelected(&row_table, static_cast<uint32_t>(num_rows),
+                                       row_ids_right.data()));
+
+  ASSERT_TRUE(row_table.offsets());
```
Comment (pitrou) [inline cpp/src/arrow/compute/row/compare_test.cc:234]: I'm not sure what's that supposed to check (offsets being "true"?). Do we want to make the test a bit more self-documenting, or perhaps add a comment?

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc (hunk 2 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_internal_avx2.cc:239]: Oh sorry, I meant in another PR.

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc (hunk 4 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_internal_avx2.cc:276]: It's just copied from the existing usage:
https://github.com/apache/arrow/pull/42188/files#diff-4dcfb2dd9add2d770dcbd924498120ab30e38186dd4bdc60708485c31074ce67L368

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc (hunk 5 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_internal_avx2.cc:268]: 1. The type of the third formal parameter of `_mm256_set1_epi32/64` is `int` so I'm just using `int` too.
2. Yeah, will do.

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc (hunk 6 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_internal_avx2.cc:258]: Yeah, will do.

File: cpp/src/arrow/compute/row/compare_test.cc (hunk 7 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_test.cc:184]: Sorry, I meant "workaround". Will update.

File: cpp/src/arrow/compute/row/compare_test.cc (hunk 8 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_test.cc:171]: What do you mean by "runtime"? I can't think of a reason why Valgrind would complain (at least ASAN didn't).

Comment (pitrou) [reply cpp/src/arrow/compute/row/compare_test.cc:171]: Sorry, I meant "run time" or execution time :-)

File: cpp/src/arrow/compute/row/compare_test.cc (hunk 9 above)
Comment (pitrou) [reply cpp/src/arrow/compute/row/compare_test.cc:234]: Yes, the `ASSERT_NE` suggestion would make this more easily understandable, thanks!

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc (hunk 3 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_internal_avx2.cc]: Addressed.

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc (hunk 5 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_internal_avx2.cc]: Addressed.

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc (hunk 6 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_internal_avx2.cc]: Addressed.

File: cpp/src/arrow/compute/row/compare_test.cc (hunk 7 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_test.cc]: Addressed.

File: cpp/src/arrow/compute/row/compare_test.cc (hunk 8 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_test.cc:171]: Ah, got it! It takes about 20s with ASAN enabled. Perhaps it will be fine with Valgrind too?

Comment (pitrou) [reply cpp/src/arrow/compute/row/compare_test.cc:171]: I should take a quick look.

Comment (pitrou) [reply cpp/src/arrow/compute/row/compare_test.cc:171]: Ok, it takes 70s locally under Valgrind. That's a bit high for a single test, I would rather disable it under Valgrind.

Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_test.cc:172]: OK. Updated to disable the test under Valgrind. Thanks for helping running in your local!

File: cpp/src/arrow/compute/row/compare_internal_avx2.cc [hunk 10]
```diff
@@ -251,6 +253,40 @@ uint32_t KeyCompare::CompareBinaryColumnToRowHelper_avx2(
...
+}
+
+template <uint32_t kScale>
+inline __m256i UnsignedOffsetSafeGather64(arrow::util::int64_for_gather_t const* base,
+                                          __m128i offset) {
+  arrow::util::int64_for_gather_t const* normalized_base =
+      base + kTwoGB / sizeof(arrow::util::int64_for_gather_t);
+  __m128i normalized_offset =
+      _mm_sub_epi32(offset, _mm_set1_epi32(static_cast<int>(kTwoGB / kScale)));
+  return _mm256_i32gather_epi64(normalized_base, normalized_offset,
```
[the remark this answers is not in the document]
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/compare_internal_avx2.cc:284]: Both intrinsics gather "several" integers based on a base address and "several" 32b offsets (`vindex`), and stores the results into a 256b register. The difference is: `_mm256_i32gather_epi32` gathers 8 32b-integers (8 * 32 = 256) at a time so 8 32b indices are used, hence the 256b `vindex`. Whereas `_mm256_i32gather_epi64` gathers 4 64b-integers at a time so 4 32b indices are used, hence the 128b `vindex`.

[the remark this answers is not in the document]
Comment (ZhangHuiGui) [reply cpp/src/arrow/compute/row/compare_internal_avx2.cc:284]: I sees. Thanks!

**Comment by zanmato1984** [comment]
I've committed another change containing only code restructures (moving, renaming, etc.) to the test, to make the test logic more clear and readable. Hope it doesn't trouble your review @pitrou . Thanks.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thanks a lot @zanmato1984 !

**Pull request merged**